### PR TITLE
Added barrier sync functionality.

### PR DESCRIPTION
**Pull request by andrewbenton**
Public API is exposed in the Barrier class.  This allows test scenarios to synchronize between themselves, and includes support for tests not making the barrier, timeouts, and distinct code paths.  Code operating under this barrier sync implementation may not, and likely will not, execute properly when mixing the barrier sync with a serialized test, however it will still run.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -81,26 +80,6 @@
             </plugin>
 
             <plugin>
-                <groupId>org.codehaus.gmaven</groupId>
```
Comment (dimaj) [inline pom.xml:84]: just wondering... what were these plugins used for and why did you remove them?

File: src/main/java/com/comcast/zucchini/AbstractZucchiniTest.java [hunk 2]
```diff
@@ -12,12 +17,13 @@
 import org.testng.annotations.AfterClass;
 import org.testng.annotations.Test;
 
-import com.google.gson.Gson;
 import com.google.gson.JsonObject;
 import com.google.gson.JsonArray;
 import com.google.gson.JsonElement;
 import com.google.gson.JsonParser;
 
+import cucumber.runtime.CucumberException;
```
Comment (dimaj) [inline src/main/java/com/comcast/zucchini/AbstractZucchiniTest.java]: this exception is not being used, please remove

File: src/main/java/com/comcast/zucchini/FlexibleBarrier.java [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+            //if the thread has not arrived or already been registered as failed, register it as failed, and stop it
+
+            if(!this.arrivedThreads.contains(tc)) { //if it didn't successfully meet the barrier
+                if(!this.azt.failedContexts.contains(tc)) { //and it wasn't already registered as failed
+                    synchronized(azt.failedContexts) { //lock the contexts
+                        if(!this.azt.failedContexts.contains(tc)) { //test again for race condition prevention
+                            this.azt.failedContexts.add(tc);
+                            this.dec();
+                            if(tc.canKill) {
+                                tc.getThread().stop();
```
Comment (dimaj) [inline src/main/java/com/comcast/zucchini/FlexibleBarrier.java:73]: Thead.stop() has been deprecated...
http://stackoverflow.com/questions/10961714/how-to-properly-stop-the-thread-in-java

File: src/main/java/com/comcast/zucchini/AbstractZucchiniTest.java [hunk 4]
```diff
@@ -31,18 +37,34 @@
...
-    static HashMap<String, JsonArray> featureSet = new HashMap<String, JsonArray>();
+    private static final Logger LOGGER = LoggerFactory.getLogger(AbstractZucchiniTest.class);
+    static Map<String, JsonArray> featureSet = new HashMap<String, JsonArray>();
 
     /* Synchronization and global variables.  DO NOT TOUCH! */
+    private static Object lock = new Object();
     private static Boolean hooked = false;
 
+    /* store the list of contexts here */
+    List<TestContext> contexts;
```
Comment (dimaj) [inline src/main/java/com/comcast/zucchini/AbstractZucchiniTest.java:48]: should these vars be private with a getter?

File: src/main/java/com/comcast/zucchini/AbstractZucchiniTest.java [hunk 5]
```diff
@@ -86,21 +140,25 @@ public void runParallel(List<TestContext> contexts) {
...
         }
 
         for(Thread t : threads) {
             try {
                 t.join();
             }
-            catch(Throwable e) {
-                logger.error(t.toString());
+            catch(InterruptedException e) {
+                LOGGER.error("ERROR on Thread.join(): {}", e);
```
Comment (dimaj) [inline src/main/java/com/comcast/zucchini/AbstractZucchiniTest.java:153]: if there was an exception while joining threads, should tests fail?

File: src/main/java/com/comcast/zucchini/AbstractZucchiniTest.java (hunk 4 above)
Comment (andrewbenton) [inline src/main/java/com/comcast/zucchini/AbstractZucchiniTest.java:48]: It's used at package-level access.  For now, it's not a part of the API, and people inheriting from AZT aren't able to access it anyway.  Providing a getter would be an unnecessary extra step.

Comment (andrewbenton) [inline src/main/java/com/comcast/zucchini/AbstractZucchiniTest.java:48]: I'll take a look to check and be sure that I'm using it though.  I'm pretty sure that I use it in sizing the barriers auto-magically.

File: src/main/java/com/comcast/zucchini/AbstractZucchiniTest.java (hunk 5 above)
Comment (andrewbenton) [inline src/main/java/com/comcast/zucchini/AbstractZucchiniTest.java:153]: That might be a good idea, thought the fail should wait until they have all joined or failed, if we choose to go that route.  We might need to add the MojoExecutionException back in as the 'throw' if we want it run later tests.  

On a side note, I don't think that the InterruptedException can make it out to this point unless the interrupted flag is set and is then allowed to propagate.  If it is raised in on the of the step definitions, it won't make it back up to this level.  I'll have to check the docs again to be sure.

File: pom.xml (hunk 1 above)
Comment (andrewbenton) [inline pom.xml:84]: Groovy integration for apache maven: https://groovy.github.io/gmaven/

File: src/main/java/com/comcast/zucchini/AbstractZucchiniTest.java (hunk 2 above)
Comment (andrewbenton) [inline src/main/java/com/comcast/zucchini/AbstractZucchiniTest.java]: Gladly :)

File: src/main/java/com/comcast/zucchini/FlexibleBarrier.java (hunk 3 above)
Comment (andrewbenton) [inline src/main/java/com/comcast/zucchini/FlexibleBarrier.java:73]: It's unfortunately necessary, given that we are trying to prevent bad code from causing Zucchini to block on the barrier sync.  The Thread.interrupt() is only useful if the code within the thread is checking for interruptions or is performing IO operations.  

The Thread.stop() method will allow us to break from bad code such as 'while(true) {}'.  If I remove this, then we can't enforce the barrier timeout. @trentontrees

Comment (trentontrees) [inline src/main/java/com/comcast/zucchini/FlexibleBarrier.java:73]: I'm fine with the approach that you have supplied here using Thread.stop().

File: src/main/java/com/comcast/zucchini/FlexibleBarrier.java [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+    synchronized void refresh() {
+        this.reset();
+        this.primary.bulkRegister(this.azt.contexts.size() - this.primary.getRegisteredParties());
+        this.secondary.bulkRegister(this.azt.contexts.size() - this.secondary.getRegisteredParties());
+    }
+
+    /**
+     * A shortcut to get the {@see TestContext} name or provide "<NULL>" it its place.
+     */
+    private static String name() {
```
Comment (dimaj) [inline src/main/java/com/comcast/zucchini/FlexibleBarrier.java]: Factor this out to a utility class as it is used / defined in multiple places. You can use `import static` in your other classes to save on typing.

File: src/main/java/com/comcast/zucchini/ZucchiniRuntime.java [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+        formatter.done();
+        formatter.close();
+        this.printSummary();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void printSummary() {
```
Comment (dimaj) [inline src/main/java/com/comcast/zucchini/ZucchiniRuntime.java]: remove all methods that just call `super.methodName`

File: src/main/java/com/comcast/zucchini/ZucchiniShutdownHook.java [hunk 8]
```diff
@@ -37,22 +39,29 @@ public void run() {
...
-            logger.error("FATAL ERROR: " + t.getMessage());
-            /* must use system.halt here, system.exit stalls */
+        catch(Exception t) {
+            LOGGER.error("FATAL ERROR: " + t.getMessage());
             Runtime.getRuntime().halt(-1);
         }
+        finally {
+            if(writer != null) {
+                try {
+                writer.close();
```
Comment (dimaj) [inline src/main/java/com/comcast/zucchini/ZucchiniShutdownHook.java]: please indent
